File: app/src/main/java/com/example/pc/basemvp/widget/SuperRecyclerView.java
package com.example.pc.basemvp.widget;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SimpleItemAnimator;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import com.example.pc.basemvp.R;
import com.example.pc.basemvp.base.recyclerView.EndlessRecyclerOnScrollListener;

public class SuperRecyclerView extends FrameLayout {

    private ProgressBar mProgressBar;
    private RecyclerView mRecyclerView;
    private RecyclerView.OnScrollListener mOnScrollListener;
    private SwipeRefreshLayout mSwipeRefreshLayout;
    private LoadDataListener mListener;

    public SuperRecyclerView(@NonNull Context context) {
        super(context);
        initView();
    }

    public SuperRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        initView();
    }

    public SuperRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs,
            int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView();
    }

    private void initView() {
        View view =
                LayoutInflater.from(getContext()).inflate(R.layout.super_recyclerview, this, false);
        mRecyclerView = view.findViewById(R.id.recyclerView);
        mProgressBar = view.findViewById(R.id.progressBar);
        mSwipeRefreshLayout = view.findViewById(R.id.swipe_refresh_layout);
        addView(view);
    }

    public void setLoadDataListener(LoadDataListener listener) {
        mListener = listener;
    }

    public void setLayoutManager(RecyclerView.LayoutManager layoutManager) {
        mRecyclerView.setLayoutManager(layoutManager);
        mOnScrollListener = new EndlessRecyclerOnScrollListener(layoutManager) {
            @Override
            public void onLoadMore(int currentPage) {
                startLoadMore();
                mListener.onLoadMore(currentPage);
            }
        };
        mRecyclerView.addOnScrollListener(mOnScrollListener);

        mSwipeRefreshLayout.setOnRefreshListener(() -> {
            setRefreshing(true);
            mListener.onRefreshData();
        });
    }

    public void setAdapter(RecyclerView.Adapter adapter) {
        mRecyclerView.setAdapter(adapter);
    }

    public void setRefreshing(boolean isRefresh) {
        mSwipeRefreshLayout.setRefreshing(isRefresh);
    }

    public void stopLoadData() {
        setRefreshing(false);
        stopLoadMore();
    }

    public void startLoadMore() {
        mRecyclerView.setOverScrollMode(OVER_SCROLL_NEVER);
        mProgressBar.setVisibility(View.VISIBLE);
    }

    public void stopLoadMore() {
        mRecyclerView.setOverScrollMode(OVER_SCROLL_ALWAYS);
        mProgressBar.setVisibility(View.GONE);
    }

    public void disableAnimateRecyclerView() {
        ((SimpleItemAnimator) mRecyclerView.getItemAnimator()).setSupportsChangeAnimations(false);
    }

    public void enableAnimateRecyclerView() {
        ((SimpleItemAnimator) mRecyclerView.getItemAnimator()).setSupportsChangeAnimations(true);
    }

    public void resetState() {
        ((EndlessRecyclerOnScrollListener) mOnScrollListener).reset();
    }

    public void release() {
        mSwipeRefreshLayout.setOnRefreshListener(null);
        mSwipeRefreshLayout = null;
        mRecyclerView.removeOnScrollListener(mOnScrollListener);
        mRecyclerView = null;
        mProgressBar = null;
    }

    public RecyclerView getRecyclerView() {
        return mRecyclerView;
    }

    public SwipeRefreshLayout getSwipeRefreshLayout() {
        return mSwipeRefreshLayout;
    }

    public ProgressBar getProgressBar() {
        return mProgressBar;
    }

    public interface LoadDataListener {
        void onLoadMore(int page);

        void onRefreshData();
    }
}

File: app/src/main/java/com/example/mvvmkoin/widget/superRecyclerView/SuperRecyclerView.kt
package com.example.mvvmkoin.widget.superRecyclerView

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.FrameLayout
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import androidx.recyclerview.widget.*
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.example.mvvmkoin.R
import com.example.mvvmkoin.util.Constant
import com.example.mvvmkoin.util.LogUtils
import com.example.mvvmkoin.util.extension.notNull
import com.example.mvvmkoin.widget.recyclerView.EndlessRecyclerOnScrollListener
import com.example.mvvmkoin.widget.recyclerView.LoadMoreAdapter

class SuperRecyclerView : FrameLayout {

    private lateinit var recyclerView: RecyclerView
    private lateinit var onScrollListener: EndlessRecyclerOnScrollListener
    private lateinit var loadMoreAdapter: LoadMoreAdapter<*>
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout

    private var isRefresh = false
    private var currentPage = Constant.PAGE_DEFAULT

    private var listener: LoadDataListener? = null

    constructor(@NonNull context: Context) : super(context) {
        initView()
    }

    constructor(@NonNull context: Context, @Nullable attrs: AttributeSet) : super(context, attrs) {
        initView()
    }

    constructor(@NonNull context: Context, @Nullable attrs: AttributeSet,
            defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initView()
    }

    private fun initView() {
        val view = LayoutInflater.from(context).inflate(R.layout.super_recyclerview, this, false)

        recyclerView = view.findViewById(R.id.super_recyclerView)
        recyclerView.setHasFixedSize(true)
        recyclerView.itemAnimator = DefaultItemAnimator()

        swipeRefreshLayout = view.findViewById(R.id.swipe_refresh_layout)
        swipeRefreshLayout.setColorSchemeColors(*colorForSmart)

        addView(view)
    }

    fun setUpView(adapter: LoadMoreAdapter<*>,
            layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(context)) {
        loadMoreAdapter = adapter
        recyclerView.adapter = adapter

        if (layoutManager is GridLayoutManager) {
            layoutManager.spanSizeLookup = getSpanSizeLookup(layoutManager)
        }
        recyclerView.layoutManager = layoutManager

        onScrollListener = object : EndlessRecyclerOnScrollListener(layoutManager) {
            override fun onLoadMore(currentPage: Int) {
                startLoadMore()
            }
        }
        recyclerView.addOnScrollListener(onScrollListener)

        swipeRefreshLayout.setOnRefreshListener { startRefreshData() }
    }

    fun startLoadMore() {
        if (loadMoreAdapter.isLoadMore || isRefresh) {
            return
        }
        currentPage++
        recyclerView.post { loadMoreAdapter.onStartLoadMore() }
        listener?.onLoadMore(currentPage)
    }

    fun stopLoadMore() {
        loadMoreAdapter.onStopLoadMore()
    }

    private fun startRefreshData() {
        if (isRefresh) return

        isRefresh = true
        currentPage = Constant.PAGE_DEFAULT

        swipeRefreshLayout.isRefreshing = true
        listener?.onRefreshData()
    }

    fun stopRefreshData() {
        if (!isRefresh) return
        isRefresh = false
        swipeRefreshLayout.isRefreshing = false
    }

    fun refreshAdapter() {
        stopLoadMore()
        resetScrollState()
    }

    fun disableAnimate() {
        (recyclerView.itemAnimator as SimpleItemAnimator).supportsChangeAnimations = false
    }

    fun enableAnimate() {
        (recyclerView.itemAnimator as SimpleItemAnimator).supportsChangeAnimations = true
    }

    fun setHasFixedSize(isFixed: Boolean) {
        recyclerView.setHasFixedSize(isFixed)
    }

    fun setEnableSwipe(isEnable: Boolean) {
        swipeRefreshLayout.isEnabled = isEnable
    }

    private fun getSpanSizeLookup(
            layoutManager: GridLayoutManager): GridLayoutManager.SpanSizeLookup {
        val spanSize = layoutManager.spanCount
        return object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                return if (loadMoreAdapter.isLoadMore
                        && position == loadMoreAdapter.bottomItemPosition()) {
                    spanSize
                } else 1
            }
        }
    }

    private fun resetScrollState() {
        onScrollListener.reset()
    }

    fun setLoadDataListener(listener: LoadDataListener) {
        this.listener = listener
    }

    interface LoadDataListener {
        fun onLoadMore(page: Int)

        fun onRefreshData()
    }

    override fun onDetachedFromWindow() {
        listener = null
        loadMoreAdapter.notNull { it.unRegisterItemClickListener() }
        swipeRefreshLayout.notNull { it.setOnRefreshListener(null) }
        onScrollListener.notNull { recyclerView.removeOnScrollListener(it) }
        LogUtils.d(TAG, Constant.RELEASED)
        super.onDetachedFromWindow()
    }

    companion object {
        private val TAG = SuperRecyclerView::class.java.simpleName
        val colorForSmart = intArrayOf(Color.rgb(30, 186, 177), Color.rgb(0xF7, 0xD2, 0x3E),
                Color.rgb(0xF7, 0xD2, 0x3E), Color.rgb(0x34, 0xA3, 0x50))
        val colorForRing = intArrayOf(Color.rgb(0xF7, 0xD2, 0x3E), Color.rgb(0xF7, 0xD2, 0x3E),
                Color.rgb(30, 186, 177), Color.rgb(0x34, 0xA3, 0x50))
    }
}
